docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,158 @@
+
+```
+$ spin login --url https://example.com
+Open the Platform device authorization URL in your browser: https://example.com/device-authorization and enter the code: tWOJCFhY
+Waiting for device authorization...
+Waiting for device authorization...
+Device authorized!
+Login successful.
+```
+
+The URL and token generated by the device authorisation flow is stored on the machine.
+
+A user may log into multiple Hippo servers.  The (URL, token) pairs for each will be stored.  That is, logging into one server does not log you out of others.
+
+> QUESTION: Or should you only be logged into one at a time?
+
+> QUESTION: Should we provide an explicit `logout` command to destroy the token?
+
+### Deploy command changes
+
+The `spin deploy` command will now act as follows:
+
+* The Hippo URL, username and password are now **optional** at the implementation level.
+* If the Hippo URL is specified:
+  * If the user is logged into that server, use the stored token for auth.  The username and password are not required and _must not be present_.
+  * If the user is not logged into that server, and a username or password is provided, follow the current behaviour.  Both username and password are required.
+  * If the user is not logged into that server, and no username or password is provided, attempt a login using the same behaviour as `spin login`.  If that succeeds, continue to deployment.  If it fails, exit.
+* If the Hippo URL is **not** specified:
+  * If a username or password is specified, error. A URL is required in this case.
+  * If the user is logged in, use the stored URL and token.
+  * If the user is **not** logged in, attempt to log them into a default URL, such as the default address of a local Fermyon platform. If that fails, print usage information and exit.
+
+> QUESTION: If a user is logged into multiple servers, should we:
+> * Default the URL to the _most recent_ login
+> * Provide a way for the user to say which URL is their default
+> * Require the user to provide a URL
+> * Provide an abstraction (e.g. name, config file path) by which the user can switch between servers
+
+> QUESTION: The last section represents a change in behaviour when no flags are provided. Are we happy with this, or should we direct people to `spin login` and not attempt it automatically?
+
+> QUESTION: Is it correct to reject a username/password if the user is logged in to the chosen server, or should we attempt username/password login in this case?
+
+> QUESTION: What happens if the user is logged in to one server and `$HIPPO_URL` is set to another?  What happens if the user is logged in but the `$HIPPO_USERNAME` is set?  (Thinking here of situations where the user is switching between a username/password server and a device-auth one.)  Should we provide a way to "log into" username/password servers, and use that stored state instead of the environment variables?
```

Comment:
Aha... your second remark makes me suspect that you have a mental model of how `HIPPO_USERNAME` and `HIPPO_PASSWORD` map onto token servers, which would explain some of your other remarks that I hadn't quite followed.  (Whereas I have been treating the username-password UI as largely distinct from the token UI.)  Let's maybe have a chat and see if we can capture and document that understanding?

Reply:
Re your first comment, I think you maybe have a mental model where any deployment is a login?  Like if I deploy to server A, that logs me into server A; then if I set `HIPPO_URL` and deploy to server B, that logs me into server B?  There is no (or less) sense of command options applying just to the current deploy command?  Have I understood this right?  Again, happy to chat and get a sense for how you see this.

Reply:
To clarify, my mental model was that:

1. Command line options (and hence EVs) affected _that one execution of the command_ and overrode defaults
2. Only an explicit login step (whether via `spin login` or prompted from `spin deploy`) would set new defaults

(This is not to argue for either model... I just realised I hadn't explicitly written mine down!)

Reply:
> Command line options (and hence EVs) affected that one execution of the command and overrode defaults

Following up from https://github.com/fermyon/spin/pull/795#discussion_r986028506. If we were to match that mental model, then there would not be a way to have these options affect a single execution because they would not exist as part of `spin deploy`.

I would argue removing the flags and moving them to `spin login`, forcing the user to log in, is the better approach. Consider the current implementation of `spin deploy` in #794:

```
$ export HIPPO_URL=https://***************
$ spin deploy
$ vim src/lib.rs
$ vim spin.toml
$ spin deploy
$ sleep 30d # wait until the session token expires
$ spin deploy
```

If the server implements Github OAuth sign-in, and EVs affect the single execution of the command, the user will be forced to enter a device code three times - twice in short succession of each other, and once more a week later. I feel this would be poor/clunky design from a user's perspective.

My mental model looks more like this:

```
$ export HIPPO_URL=https://***************
$ spin login
$ spin deploy
$ vim src/lib.rs
$ vim spin.toml
$ spin deploy
$ sleep 30d # wait until the session token expires
$ spin deploy
```

The user will be asked to enter a device code once (`spin login`). After the session token expires, `spin deploy` will check the session token and notice it has expired. It will either error out ("Your session has expired. Please use `spin login` to log in"), or it will invoke `spin login` before continuing with `spin deploy`. Kinda like how `helm upgrade --install` [would invoke `helm install`](https://github.com/helm/helm/blob/45161a5366ca9dd5de2a070336d4bdc191448d6b/cmd/helm/upgrade.go#L95-L126) if it was the first time the user was installing a chart.